Update browser output for front end pipeline

KatalinaBobowik opened the pull request:
These updates were made to address the GitHub issues outlined [here](https://github.com/populationgenomics/tob-wgs/issues/97). The association output has been modified so that it is easier to read into the front-end workflow, as well as reduce computational cost by combining into one script. *Note some code may change once Hail Batch is updated tomorrow morning (see [this post](https://centrepopgen.slack.com/archives/C018KFBCR1C/p1657589184890679)). 

@daniaki and @danielreti , I've addressed the changes in the doc, but please let me know if there's something that should be changed or added. I also haven't added in the `LD information` section, as this will be run on the unique union set of significant SNPs across all cell types after running the association analysis on the whole genome.

Commit:
Add in association effect on gene expression

Commit:
test appending parquet file using 'ab'

Commit:
Try exporting table to spark

Commit:
Save each file to a parquet

Commit:
Add in log cpm to get_association_effect_data()

Commit:
Add dataproc step to convert hail table to parquet

Commit:
Add in to_spark() in order to convert to spark before conversion to parquet

Commit:
Use output_path rather than AnyPath

As per error messgae here https://batch.hail.populationgenomics.org.au/batches/77052/jobs/4

Commit:
Add in global bp and save hail tabel in correct format

Commit:
Change incorrectly named file

Commit:
Pass input path as a click parameter

This is a workaround; see https://centrepopgen.slack.com/archives/C030X7WGFCL/p1656479712892089?thread_ts=1656327147.623059&cid=C030X7WGFCL

Commit:
Correctly read in input files in dataproc script

Commit:
Add in invocation to run function

Commit:
Update conditional analysis to have same specs as specified by front end

Specs listed here: https://github.com/populationgenomics/ancestry/blob/main/scripts/hail_batch/calculate_ld/calculate_ld.py

Commit:
Update README

Commit:
Add argument to ensure cluster runs after spearman correlation

Commit:
Remove 'depends_on=j' on to debug

Commit:
Add 'depends_on' after job is created

Commit:
Add in 'depends_on' as list syntax

Commit:
Remove depends_on argument to debug

Commit:
Make dataproc cluster dependent on spearman correlation output

Commit:
Save eqtl_effect data as csv instead of hail table

Commit:
Save files to temporary path

Commit:
Save association effect file using AnyPath

Commit:
Add in job to calculate ld

Commit:
Remove comment line

Commit:
Feed in result_second to calculate ld function

Commit:
initialise hail

Commit:
Add in print statements to debug

Commit:
Split merge df function

Split function so that output returns a merged df (not output as string) so that this can be fed into the calculate ld function. Then create a new function to export reuslts to a string

Commit:
COmment out functions to debug

Debugging after error TypeError: 'PythonResult' object is not callable. See https://batch.hail.populationgenomics.org.au/batches/78041/jobs/1

Commit:
Change name in mt to global_bp for consistency

Commit:
Test running ld in dataproc

Commit:
Remove unnecessary calculate_ld function in script

Commit:
Fix dependency to be a 'new_python_job'

Commit:
Use variable instead of full path for filename

Commit:
Create variable name for input mt

Commit:
Update incorrect 'ht' to 'mt'

Commit:
Update README

Commit:
Update all instances of snpid to snp_id

As per Daniel Espositos request (this is the data format expected by the web UI_

Commit:
Remove all instances of 'chr' in get_genotype_df()

Commit:
Add in 'chr' when assigning chromosome

If 'chr' is not included, global_bp fails ('chr' indicates GrCh38 format)

Commit:
Make merged_sig_snps a string

This might potentially solve the error 'TypeError: unsupported operand type(s) for -: 'NoneType' and 'int''; see https://batch.hail.populationgenomics.org.au/batches/78699/jobs/4

Commit:
Remove to_string() requirement from merged_sig_snps

Commit:
Fix issue where qvalue only gets first entry

Commit:
Add in print statements to debug

Commit:
Comment out lines which might be failing for debugging

Commit:
Uncomment out text (debug)

Commit:
Uncommenting code to debug

Commit:
Uncomment more lines to debug

Commit:
Add print statements to debug

Commit:
Remove desnify to test why script is breaking

Print statements (debugging) points to this being the issue

Commit:
get output of ld matrix to have global_bp rather than row idx

Commit:
Add in VEP annotation

Commit:
Fix pylint errors

Commit:
Only write header to file if it doesn't exist

Commit:
Save files as tab-separated and gzipped

Commit:
Test saving files using gzip command

Previous files were not openening and getting the error: 'Not a gzipped file (b'sn')'

Commit:
Add densify back in to test whether it is fixed

Commit:
Remove densify again to test if that solves the issue

Commit:
Write out header to csv file if file doesn't exist

Test another way of saving teh file, as suggested by Daniel Esposito

Commit:
Only get upper triang;e for LD calculation

Commit:
Revert back to csv file (unzipped)

Commit:
Update csv filename to end in csv rather than tsz.gz

Commit:
Try writing out file as gzip with path.open('wb')

As per Leo's recommendation: https://centrepopgen.slack.com/archives/C030X7WGFCL/p1657520711610929?thread_ts=1657520513.139129&cid=C030X7WGFCL

Commit:
Save eqtl_effect file with no header if it already exists

Commit:
Fix linter errors

Commit:
Make output of merged_df tab separated

Commit:
Read in significant SNPs as tsv

Commit:
Update README

katiedelange commented on scripts/eqtl_hail_batch/calculate_ld.py:
```diff
@@ -0,0 +1,64 @@
+    )
+    t = hl.Table.from_pandas(significant_snps_df)
+    # only keep rows whose FDR is < 0.05
+    # t = t.filter(t.fdr < 0.05)
```
Should this be uncommented? Alternatively, if it's no longer needed because you only load significant SNPs into the dataframe above, can you expand the significant SNPs comment to include this threshold information (FDR < 0.05) before removing the unnecessary lines.

KatalinaBobowik replied:
You're 100% correct- this should be uncommented (I think that's a holdover from testing).

katiedelange commented on scripts/eqtl_hail_batch/calculate_ld.py:
```diff
@@ -0,0 +1,64 @@
+    # only calculate the upper triangle 
+    ld = ld.sparsify_triangle()
+    table = ld.entries()
+    # filter out entries with an LD score less than 0.2
```
Given we are only dealing with significant SNPs within small loci, is this filtering necessary to maintain reasonable table sizes, or could we drop it? Two significant SNPs with such low LD between them might occasionally be interesting (depending on the shape of the association signal in the locus, where these variants sit, and the relationship between them) and it would be nice to be able to display the actual LD between them if possible.

If we need to keep the filtering, then we at least need to be clear in the browser when no value = very low LD, versus no value = not tested.

KatalinaBobowik replied:
I'm happy to remove it, only previously the majority of rows were taken up by very low LD values. in GTEx, the minimum they use is 0.1, but in a data meeting I think we settled on 0.2. Very happy to change this though and not use any filter, as, like you said, we're only dealing with significant SNPs now. Regardless, if we set some minimum value, I agree that we should make the clear in the browser. 
Just a side note, the `calculate_ld.py` script will need further refining as I test it once all the significant SNPs come in, so this can be a conversation we have with the Daniels, or I can test and see how many results I actually get and how large the output matrix is. @daniaki or @danielreti , do you have any thoughts?

daniaki replied:
Testing without the filter first sounds like a good idea, then if we need to slim the results down we can add a filter of 0.1 like GTEx.

KatalinaBobowik replied:
Sounds great, I'll test that one first :)

katiedelange commented on scripts/eqtl_hail_batch/calculate_ld.py:
```diff
@@ -0,0 +1,64 @@
+    # turn matrix into table and save, in order to reference row idx
+    mt_path = f'{input_path}significant_snps.mt'
+    mt.write(mt_path)
+    # perform ld calculation
```
Nitpick: can you expand the comment to explain that we are calculating LD for a 2MB window - radius=2e-6 is a bit of a confusing parameter name.

KatalinaBobowik replied:
updated :)

katiedelange commented on scripts/eqtl_hail_batch/conditional_analysis.py:
```diff
@@ -134,29 +134,34 @@ def get_genotype_df(filtered_mt_path, residual_df, gene_snp_test_df):
 
     genotype_df = t.to_pandas(flatten=True)
     genotype_df.rename({'onek1k_id': 'sampleid'}, axis=1, inplace=True)
+    # chromosome must be turned into a number solely
```
Is this change just for cosmetics when displaying, or because we are storing chromosome as a numeric column to make the database more efficient etc? If we want to treat things as numeric, there may need to be some additional transformation here to handle text chromosomes (X, Y, MT).

Note that using numeric encoding also opens a can of worms, as there aren't consistent maps for these chromosomes to numbers - i.e. Plink uses X=23, Y=24, XY=25, MT=26 vs most other software uses X=23, Y=24, MT=25. So we need to make an explicit decision on these encodings and ensure they are used consistently everywhere.

KatalinaBobowik replied:
This change is to address the specifications outlined in the [GitHub issues doc](https://github.com/populationgenomics/tob-wgs/issues/97). @daniaki and @danielreti might know best about why the change is preferred, and if it's not, I can change it back (it would actually make certain parts of my script cleaner, since hail needs the 'chr' prefix to designate it's GRCh38). However, in saying that, everything here is a string representation of the number, so non-numeric.

daniaki replied:
This format was to match the output I got from Joseph's group, but it shouldn't cause any major issues on our end if you change it back. Feel free to change it back particularly if it makes other parts of the script cleaner.

katiedelange commented on scripts/eqtl_hail_batch/generate_eqtl_spearman.py:
```diff
@@ -195,7 +202,7 @@ def prepare_genotype_info(keys_path, expression_path):
         log_expression_df = get_log_expression(expression_df)
         mt = hl.read_matrix_table(TOB_WGS)
-        mt = hl.experimental.densify(mt)
+        # mt = hl.experimental.densify(mt)
```
Can this be restored now, following [this fix](https://github.com/populationgenomics/hail/pull/206)?

KatalinaBobowik replied:
yes, sorry that as well is a holdover. I have now uncommented this

katiedelange commented on scripts/eqtl_hail_batch/generate_eqtl_spearman.py:
```diff
@@ -388,6 +404,62 @@ def run_spearman_correlation_scatter(
+            mean_val = group[gene].mean()
+            q1, median_val, q3 = group[gene].quantile([0.25, 0.5, 0.75])
+            iqr = q3 - q1
+            # not sure what this value is doing here
```
What is this comment referring to - if the interquartile ranges aren't used for the violin plots, do we need to calculate them and include them in the struct?

daniaki replied:
We were planning to use `iqr_min` and `iqr_max` to draw boxplot whiskers, but I've just realised that whiskers are drawn up/down to the largest/smallest value within `iqr_max/iqr_min` and not those values themself. Is it ok to omit the calculation of `iqr_min/iqr_max` and use the `min/max` as the whisker end points instead?

KatalinaBobowik replied:
Sorry, that's a very disconcerting comment to have in code 🤦‍♀️ As @daniaki said, that's a holdover from previously, when I wasn't sure what the calculation was used for, as I got it from the GitHub issues doc. What I found out is that that calculation is used to identify outliers (hence, the 1.5 constant). @daniaki no problem, so since the min and max value are already there, just keep delete the `iqr_min` and `iqr_max`? I'll go ahead and remove them from the code and the data struct

Commit:
Address reviewer comments as in https://github.com/populationgenomics/tob-wgs/pull/98

Commit:
Add celltype into batch name to ensure unique ID

Commit:
Add 'chr' prefix back into all instances of chromosome

Commit:
Remove all LD filters in first test run

Commit:
Update output file extension in wrapper script

Commit:
separate files by '/'

KatalinaBobowik commented:
@katiedelange @daniaki @danielreti I removed the filtering step on the ld calculation and reverted the chromosome field in the association output back to the 'chr' prefix. Tests are also all running sucessfully (including the wrapper script 🎉). Let me know what you think.

katiedelange reviewed (approved):
Looks great, thanks Kat! 

Only one general comment for all of us to consider more broadly: several of these changes are to adjust whether CSV or TSV is used as input/output. Going forward it would probably be good if we could all agree on a consistent format that we will use (as much as possible) to pass this sort of data back and forth.

daniaki commented:
Thanks Kat! In reference to @katiedelange's point above, my preference goes to TSV (gzipped to save space for large files) as outputs to avoid issues when commas appear in column values.

KatalinaBobowik commented:
great, thanks @daniaki. Maybe we can set that as the default then? I'll make sure to use TSV going forward
